src/client/QXmppAtmManager.cpp

```diff
@@ -0,0 +1,303 @@
+}
+
+void QXmppAtmManager::handleMessageReceived(const QXmppMessage &message)
+{
+    if (const auto trustMessageElement = message.trustMessageElement()) {
+        const bool isTrustMessageIntendedForAtm = trustMessageElement->usage() == ns_atm;
+
+        if (isTrustMessageIntendedForAtm) {
+            // TODO: Replace by real key ID from OMEMO message
+            const QString senderKeyId = "623548d3835c6d33ef5cb680f7944ef381cf712bf23a0119dabe5c4f252cd02f";
+            const QString senderJid = QXmppUtils::jidToBareJid(message.from());
+            const QString ownJid = client()->configuration().jidBare();
+            const bool isOwnTrustMessage = senderJid == ownJid;
+
+            for (const auto &keyOwner : trustMessageElement->keyOwners()) {
+                const QString keyOwnerJid = keyOwner.jid();
+
+                // Trust messages from an own endpoint may authenticate or
+                // distrust the keys of all own endpoints and endpoints of contacts.
+                // Whereas trust messages from endpoints of contacts may only
+                // authenticate or distrust the keys of their own endpoints.
+                const bool isSenderQualifiedForTrustDecisions = isOwnTrustMessage || senderJid == keyOwnerJid;
+
+                if (isSenderQualifiedForTrustDecisions) {
+                    const auto encryption = trustMessageElement->encryption();
+                    const auto idsOfKeysBeingAuthenticated = keyOwner.identifiersOfTrustedKeys();
+                    const auto idsOfKeysBeingDistrusted = keyOwner.identifiersOfDistrustedKeys();
+                    const auto isSenderKeyAuthenticated = trustStorage.isKeyAuthenticated(encryption, keyOwnerJid, senderKeyId);
+
+                    if (isSenderKeyAuthenticated) {
+                        authenticate(encryption, keyOwnerJid, idsOfKeysBeingAuthenticated);
+                        distrust(encryption, keyOwnerJid, idsOfKeysBeingDistrusted);
+                    } else {
+                        // Store key IDs received from endpoints with not yet
+                        // authenticated keys for later authentication or
+                        // distrusting.
+                        trustStorage.addKeyIdsForLaterTrustActions(encryption, keyOwnerJid, idsOfKeysBeingAuthenticated, senderKeyId, true);
+                        trustStorage.addKeyIdsForLaterTrustActions(encryption, keyOwnerJid, idsOfKeysBeingDistrusted, senderKeyId, false);
+                    }
+                }
+            }
+        }
+    }
+}
+
+///
+/// Authenticates keys automatically by the content of a trust message.
+///
+/// \param encryption namespace of the encryption protocol that uses the keys
+/// \param keyOwnerJid JID of the key owner
+/// \param keyIds IDs of the keys being authenticated
+/// \param isTrustMessageBeingSent true for sending a trust message after authenticating
```

Comment:
This sounds like the trust message somehow has already been sent, not like a call to send a trust message. Simply `sendTrustMessage` would be much better I think.